syntax = "proto3";

package ddscontrol;

// The DDS control service definition.
service DDSControl {
    // Initialize
    rpc Initialize (InitializeRequest) returns (GeneralReply) {}
    // Configure run
    rpc ConfigureRun (ConfigureRunRequest) returns (GeneralReply) {}
    // Start
    rpc Start (StartRequest) returns (GeneralReply) {}
    // Stop
    rpc Stop (StopRequest) returns (GeneralReply) {}
    // Terminate
    rpc Terminate (TerminateRequest) returns (GeneralReply) {}
}

// Request status
enum ReplyStatus {
    UNKNOWN = 0;
    SUCCESS = 1;
    ERROR = 2;
}

// General error
message Error {
    string msg = 1;
    int32 code = 2;
}

// General reply to requests
message GeneralReply {
    string msg = 1;
    ReplyStatus status = 2;
    Error error = 3;
}

// Initialize request
message InitializeRequest {
    int32 numWorkers = 1;
    string topoFile = 2;
}

// Configure run request
message ConfigureRunRequest {
    // TODO: Add request parameters here
}

// Start request
message StartRequest {
    // TODO: Add request parameters here
}

// Stop request
message StopRequest {
    // TODO: Add request parameters here
}

// Terminate request
message TerminateRequest {
    // TODO: Add request parameters here
}
